### PR TITLE
Implementation of flow dump for DPDK: this is an initial integration …

### PR DESCRIPTION
**Pull request by Koushik-kopanati**
Implementation of flow dump for DPDK
    This feature retrieves the p4 tables from p4 ovs

Signed-off-by: Koushik kopanati <koushik.kopanati@intel.com>

### CODE REVIEW

**Comment by ipdksw** [comment]
Result:FAILURE JenkinsConsole:https://s3.ipdk.io/ovs/PR-77-1/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/04f162f00da1b52feeb37a43b8f6347ea232381a

**Comment by Koushik-kopanati** [comment]
example of output 

[root@localhost simple_l3]# sudo ovs-p4ctl set-pipe br0 simple_l3.pb.bin p4Info.txt
[root@localhost simple_l3]# sudo ovs-p4ctl add-entry br0 ingress.ipv4_host "hdr.ipv4.dst_addr=1.1.1.1,action=ingress.send(0)"
[root@localhost simple_l3]# sudo ovs-p4ctl add-entry br0 ingress.ipv4_host "hdr.ipv4.dst_addr=2.2.2.2,action=ingress.send(1)"
[root@localhost simple_l3]# sudo ovs-p4ctl add-entry br0 ingress.ipv4_host "hdr.ipv4.dst_addr=2.2.3.3,action=ingress.send(2)"
[root@localhost simple_l3]# sudo ovs-p4ctl dump-entries br0
Table entries for bridge br0:
  table=ingress.ipv4_host priority=0 hdr.ipv4.dst_addr=0x01010101 actions=ingress.send(port=0x00000000)
  table=ingress.ipv4_host priority=0 hdr.ipv4.dst_addr=0x02020202 actions=ingress.send(port=0x00000001)
  table=ingress.ipv4_host priority=0 hdr.ipv4.dst_addr=0x02020303 actions=ingress.send(port=0x00000002)
Error: Could not find id 0 of type actions
[root@localhost simple_l3]#

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-77-2/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/02421bbf8ff27f1706ea0fcae76cd47abbf78968

File: external/PATCH-01-STRATUM [hunk 1]
```diff
@@ -2876,10 +2875,15 @@ index 9f4bb3a9..e0154d94 100644
 -  if (table_type == bfrt::BfRtTable::TableType::METER ||
 -      table_type == bfrt::BfRtTable::TableType::COUNTER) {
 +  const tdi::Device *device = nullptr;
-+  tdi::DevMgr::getInstance().deviceGet(0, &device);
++  tdi::DevMgr::getInstance().deviceGet(0 , &device);
```
Comment (5abeel) [inline external/PATCH-01-STRATUM]: dont need the space after 0

File: external/PATCH-01-STRATUM [hunk 2]
```diff
@@ -2876,10 +2875,15 @@ index 9f4bb3a9..e0154d94 100644
...
 +  std::unique_ptr<tdi::TableKey> table_key;
 +  std::unique_ptr<tdi::TableData> table_data;
 +  tdi_status_t tdi_status;
++
++  //const tdi::Device *device = nullptr;
```
Comment (5abeel) [inline external/PATCH-01-STRATUM:2920]: Remove to clean up unused code

File: external/PATCH-01-STRATUM [hunk 3]
```diff
@@ -2901,13 +2905,19 @@ index 9f4bb3a9..e0154d94 100644
...
-+        *flags, &entries);
++        *flags, &table_size);
++    entries = table_size;
    }
  
    table_keys->resize(0);
-@@ -507,14 +540,21 @@ template <typename T>
+   table_datums->resize(0);
+-  if (entries == 0) return ::util::OkStatus();
++ if (entries == 0) {
```
Comment (5abeel) [inline external/PATCH-01-STRATUM:2964]: We have two checks for same condition here. Remove the second one

File: external/PATCH-01-STRATUM [hunk 4]
```diff
@@ -2928,17 +2938,22 @@ index 9f4bb3a9..e0154d94 100644
...
 +    }
  
      table_keys->push_back(std::move(table_key));
      table_datums->push_back(std::move(table_data));
-@@ -523,18 +563,27 @@ template <typename T>
- 
+   }
+-  if (entries == 1) return ::util::OkStatus();
+-
++  if (entries == 1) {
```
Comment (5abeel) [inline external/PATCH-01-STRATUM:3014]: Same comment as above. Cleanup. Remove duplicate conditional checks

File: external/PATCH-01-STRATUM [hunk 5]
```diff
@@ -3829,32 +3845,29 @@ index 9f4bb3a9..e0154d94 100644
 +}
 +
 +::util::Status TdiSdeWrapper::AddPort(int device, int port, uint64 speed_bps,
-                                      FecMode fec_mode) {
++                                     FecMode fec_mode) {
 +  auto port_attrs = absl::make_unique<port_attributes_t>();
 +#ifdef P4SDE_DPDK_TARGET_STUB
-   ASSIGN_OR_RETURN(auto bf_speed, PortSpeedHalToBf(speed_bps));
-   ASSIGN_OR_RETURN(auto bf_fec_mode, FecModeHalToBf(fec_mode, speed_bps));
--  RETURN_IF_BFRT_ERROR(bf_pal_port_add(static_cast<bf_dev_id_t>(device),
```
Comment (5abeel) [inline external/PATCH-01-STRATUM:3834]: Why is this being removed?

File: external/PATCH-01-STRATUM [hunk 6]
```diff
@@ -6677,7 +6693,10 @@ index 9f4bb3a9..e0154d94 100644
      uint32 table_id, const TableKeyInterface* table_key,
      TableDataInterface* table_data) {
    ::absl::ReaderMutexLock l(&data_lock_);
-@@ -3106,35 +3736,47 @@ namespace {
++  LOG(ERROR) << "sample-testing-2 entering get table entries";
```
Comment (5abeel) [inline external/PATCH-01-STRATUM:7546]: Remove log

File: external/PATCH-01-STRATUM (hunk 5 above)
Comment (5abeel) [reply external/PATCH-01-STRATUM:3834]: Please check if the new patch file has this line present -- maybe the lines have just shifted in patch file and diff cannot show properly.

**Review by 5abeel — CHANGES_REQUESTED** [review]

File: external/PATCH-01-STRATUM (hunk 3 above)
Comment (Koushik-kopanati) [reply external/PATCH-01-STRATUM:2964]: I guess the line 2917 is modified to new one, anyways I will revert it to original one

File: external/PATCH-01-STRATUM (hunk 5 above)
Comment (Koushik-kopanati) [reply external/PATCH-01-STRATUM:3834]: These lines are present in Patch file

File: external/PATCH-01-STRATUM [hunk 7]
```diff
@@ -2901,14 +2903,18 @@ index 9f4bb3a9..e0154d94 100644
...
-+        *flags, &entries);
++        *flags, &table_size);
++    entries = table_size;
    }
  
    table_keys->resize(0);
-@@ -507,14 +540,21 @@ template <typename T>
- 
+   table_datums->resize(0);
+-  if (entries == 0) return ::util::OkStatus();
```
Comment (5abeel) [inline external/PATCH-01-STRATUM:2962]: Duplicate. Remove this

**Review by 5abeel — CHANGES_REQUESTED** [review]

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-77-3/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/dcc9dc28bda46fcc083b9e4fa6e8b6f3707f9abe

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-77-4/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/fa5dc410c03905393c565feb901228286b153f5a

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-77-5/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/fa5dc410c03905393c565feb901228286b153f5a

**Review by 5abeel — APPROVED** [review]

File: external/PATCH-01-STRATUM [hunk 8]
```diff
@@ -14,7 +14,6 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
-
```
Comment (nupuruttarwar) [inline external/PATCH-01-STRATUM:14]: maintain 1 line between license header and git diff

File: external/PATCH-01-STRATUM [hunk 9]
```diff
@@ -2928,17 +2933,19 @@ index 9f4bb3a9..e0154d94 100644
...
 +    if (tdi_status != TDI_SUCCESS) {
 +       goto internal_error;
 +    }
  
      table_keys->push_back(std::move(table_key));
      table_datums->push_back(std::move(table_data));
-@@ -523,18 +563,27 @@ template <typename T>
- 
+   }
+   if (entries == 1) return ::util::OkStatus();
```
Comment (nupuruttarwar) [inline external/PATCH-01-STRATUM:3019]: What does entry ==1 mean here?

File: external/PATCH-01-STRATUM [hunk 10]
```diff
@@ -3829,32 +3837,29 @@ index 9f4bb3a9..e0154d94 100644
...
 +  static int port_in;
 +  static int port_out;
 +
 +  auto port_attrs = absl::make_unique<port_attributes_t>();
 +#ifdef P4SDE_DPDK_TARGET_STUB
-+  ASSIGN_OR_RETURN(auto bf_speed, PortSpeedHalToBf(speed_bps));
-+  ASSIGN_OR_RETURN(auto bf_fec_mode, FecModeHalToBf(fec_mode, speed_bps));
+   ASSIGN_OR_RETURN(auto bf_speed, PortSpeedHalToBf(speed_bps));
+   ASSIGN_OR_RETURN(auto bf_fec_mode, FecModeHalToBf(fec_mode, speed_bps));
+-  RETURN_IF_BFRT_ERROR(bf_pal_port_add(static_cast<bf_dev_id_t>(device),
```
Comment (nupuruttarwar) [inline external/PATCH-01-STRATUM:4091]: why is this being removed?

File: external/PATCH-01-STRATUM [hunk 11]
```diff
@@ -3912,9 +3918,11 @@ index 9f4bb3a9..e0154d94 100644
 +      RETURN_IF_TDI_ERROR(bf_status);
 +  }
 +
-+  return ::util::OkStatus();
-+}
-+
+   return ::util::OkStatus();
+ }
+ 
+-::util::Status BfSdeWrapper::DeletePort(int device, int port) {
```
Comment (nupuruttarwar) [inline external/PATCH-01-STRATUM:4153]: I see lot of changes in stratum patch around Add and DeletePort and BF vs TDI. Please ensure that you are applying your patch on latest version

**Review by nupuruttarwar — CHANGES_REQUESTED** [review]

File: external/PATCH-01-STRATUM (hunk 9 above)
Comment (Koushik-kopanati) [reply external/PATCH-01-STRATUM:3019]: its the existing code, just the line number changed, its checking for first table entry and then trys to fetch other entries if any

File: external/PATCH-01-STRATUM (hunk 10 above)
Comment (Koushik-kopanati) [reply external/PATCH-01-STRATUM:4091]: I didnt removed this or modified the original file, git diff generated this change

File: external/PATCH-01-STRATUM (hunk 11 above)
Comment (Koushik-kopanati) [reply external/PATCH-01-STRATUM:4153]: I just checked it I don't see anything after doing git pull

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-77-6/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/e55a30a29e0b7958eaba57ae78a43e80533b22f8

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-77-7/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/e55a30a29e0b7958eaba57ae78a43e80533b22f8

**Review by nupuruttarwar — APPROVED** [review]
LGTM

**Comment by ipdksw** [comment]
Result:FAILURE JenkinsConsole:https://s3.ipdk.io/ovs/PR-77-8/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/0f0717c82c8bb70a61952a7efddc18bcf77c33f7

**Comment by ipdksw** [comment]
Result:FAILURE JenkinsConsole:https://s3.ipdk.io/ovs/PR-77-9/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/25cf16a062949c47f8914d425f44ebec0b04c638

**Comment by ipdksw** [comment]
Result:FAILURE JenkinsConsole:https://s3.ipdk.io/ovs/PR-77-10/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/25cf16a062949c47f8914d425f44ebec0b04c638

**Comment by ipdksw** [comment]
Result:FAILURE JenkinsConsole:https://s3.ipdk.io/ovs/PR-77-11/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/b0b78309b97d9b096b600b0299e70839760ec0e7

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-77-12/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/b0b78309b97d9b096b600b0299e70839760ec0e7

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-77-13/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/b0b78309b97d9b096b600b0299e70839760ec0e7

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-77-14/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/93936b92352e4690e190073278b33a17d410e02c

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-77-15/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/93936b92352e4690e190073278b33a17d410e02c

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-77-16/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/04f7e40b2b6faddfed930d00fd8b860c548d9a69

**Pull request closed without merging**